src/transport/PointToPointBroker.cpp

```diff
@@ -502,6 +519,87 @@ std::set<int> PointToPointBroker::getIdxsRegisteredForGroup(int groupId)
+        if (currentGroupId != NO_CURRENT_GROUP_ID) {
+            SPDLOG_WARN(
+              "Changing the current group Id in PTP broker ({} -> {})",
+              currentGroupId,
+              groupId);
+        }
+        currentGroupId = groupId;
+        int groupSize = getIdxsRegisteredForGroup(groupId).size();
+        // We assume that group indexes will always go from 0 to `groupSize - 1`
+        // We also want to start counting from one. This way, an unset sequence
+        // number (0 by default in protobuf) will indicate that it is unset.
+        if (sentMsgCount != nullptr) {
+            sentMsgCount->clear();
+            sentMsgCount->resize(groupSize, 1);
+        } else {
+            sentMsgCount = std::make_unique<std::vector<int>>(groupSize, 1);
+        }
+    }
+
+    assert(recvIdx < sentMsgCount->size());
+    msg.set_seqnum(sentMsgCount->at(recvIdx)++);
+}
+
+int PointToPointBroker::getExpectedSeqNum(int groupId, int sendIdx)
+{
+    if (groupId != currentGroupId) {
+        if (currentGroupId != NO_CURRENT_GROUP_ID) {
+            SPDLOG_WARN(
+              "Changing the current group Id in PTP broker ({} -> {})",
+              currentGroupId,
+              groupId);
+        }
+        currentGroupId = groupId;
+        int groupSize = getIdxsRegisteredForGroup(groupId).size();
+        // We assume that group indexes will always go from 0 to `groupSize - 1`
+        // We also want to start counting from one. This way, an unset sequence
+        // number (0 by default in protobuf) will indicate that it is unset.
+        if (recvMsgCount != nullptr) {
+            recvMsgCount->clear();
+            recvMsgCount->resize(groupSize, 1);
+        } else {
+            recvMsgCount = std::make_unique<std::vector<int>>(groupSize, 1);
+        }
+    }
+
+    assert(sendIdx < recvMsgCount->size());
+    return recvMsgCount->at(sendIdx);
+}
+
+void PointToPointBroker::incrementRecvMsgCount(int groupId, int sendIdx)
+{
+    if (groupId != currentGroupId) {
+        if (currentGroupId != NO_CURRENT_GROUP_ID) {
+            SPDLOG_WARN(
+              "Changing the current group Id in PTP broker ({} -> {})",
+              currentGroupId,
+              groupId);
+        }
+        currentGroupId = groupId;
+        int groupSize = getIdxsRegisteredForGroup(groupId).size();
+        // We assume that group indexes will always go from 0 to `groupSize - 1`
+        // We also want to start counting from one. This way, an unset sequence
+        // number (0 by default in protobuf) will indicate that it is unset.
+        if (recvMsgCount != nullptr) {
+            recvMsgCount->clear();
+            recvMsgCount->resize(groupSize, 1);
+        } else {
+            recvMsgCount = std::make_unique<std::vector<int>>(groupSize, 1);
+        }
+    }
+
+    assert(sendIdx < recvMsgCount->size());
+    recvMsgCount->at(sendIdx)++;
+}
+
 void PointToPointBroker::sendMessage(int groupId,
```

Comment:
Proposed change: header message to contain sequence number:

- Add `int` to `HEADER_MSG_SIZE` at the end.
- Set to `-1` when not set.
- Then add to the `transport::Message` class alongside the `size`, set on receive.
- Update this `PointToPointBroker::sendMessage` to take a sequence number too, thread that through to the message endpoint `sendMessage`.

Reply:
Note that if we include the sequence number in the message header, there's no need to modify the `PointToPointMessage` protobuf object, as we can just add the sequence number in the actual message.